### PR TITLE
Extend exports discovery to include omitted export specification.

### PR DESCRIPTION
**Pull request by ExternalReality**
All defined functions in a module should be found by haskell-emacs when
export specification is absent from module header.

### CODE REVIEW

File: HaskellEmacs.hs [hunk 1]
```diff
@@ -8,17 +9,20 @@ import           Control.Arrow
...
 import qualified Data.Attoparsec.ByteString.Lazy  as A
 import qualified Data.ByteString.Lazy.Char8       as B hiding (length)
 import qualified Data.ByteString.Lazy.UTF8        as B (length)
 import qualified Data.Map                         as M
-import           Data.Maybe                       (fromJust,fromMaybe)
-import           Data.Monoid                      ((<>))
-import           Data.Text                        (Text)
 import qualified Data.Text                        as T
-import           System.IO                        (hFlush, stdout)
+import qualified Language.Haskell.Exts.Syntax     as S (Name(Ident, Symbol))
```
Comment (ExternalReality) [inline HaskellEmacs.hs:26]: Please forgive the unsolicited sort of the import lines

File: HaskellEmacs.hs [hunk 2]
```diff
@@ -135,17 +139,43 @@ allExports :: [Text] -> (Text, [Text])
...
+
+exportFunction :: ExportSpec -> Maybe Name
+exportFunction (EVar _ qname)      = unQalifiedName qname
+exportFunction (EModuleContents _) = Nothing
+exportFunction _                   = Nothing
+
+unQalifiedName :: QName -> Maybe Name
+unQalifiedName (Qual _ name) = Just name
+unQalifiedName (UnQual name) = Just name
+unQalifiedName _             = Nothing
```
Comment (ExternalReality) [inline HaskellEmacs.hs:196]: Will only use SYB stlye generic programming if boiler plate starts to get completely out of control.

File: HaskellEmacs.hs (hunk 1 above)
Comment (knupfer) [inline HaskellEmacs.hs:26]: No problem (they were already sorted (according to stylish-haskell, alphabetically and ignoring `qualified'))

File: HaskellEmacs.hs (hunk 2 above)
Comment (knupfer) [inline HaskellEmacs.hs:196]: The boiler plate is no problem.  But it seems that it is broken in some cases.  For example on a clean haskell-fun directory, it won't compile (it splices "main" into the import list).  Note that you have to apply a bugfix from master first (just a oneliner, the bug was introduced due to the macro related refactoring). With various modules put in haskell-fun, it doesn't even do that and is stuck on the first stage of compiling. I'll look into that in the coming two days. Perhaps it is related with the macro based refactoring and you just rebased your commits before merging?

Anyway thanks for the pull request, it looks promising

File: HaskellEmacs.hs (hunk 1 above)
Comment (knupfer) [inline HaskellEmacs.hs:26]: By the way, you should consider adding highlighting for trailing whitespace to emacs.

Comment (knupfer) [inline HaskellEmacs.hs:26]: Besides, I recommend flycheck with haskell-ghc as a checker, you've got redundant parentheses (as well in the last PR) and a unused binding ("loc").

**Comment by ExternalReality** [comment]
All good catches, will fix.

**Comment by ExternalReality** [comment]
Catches were fixed. Please advise.

**Comment by knupfer** [comment]
Thanks for your aide! I'll review it untill the weekend.

**Comment by knupfer** [comment]
I've merged, with some modifications, it finally worked on my side. There is only one problem I can see. Your blog example won't work because of the language pragma.  Basically, it is now much more robust but can't handle pragmas...

**Comment by knupfer** [comment]
I've fixed the language issue

**Comment by ExternalReality** [comment]
Hey Knupfer, that you for handling that. Its great that it all worked out.

**Comment by knupfer** [comment]
It's great to work with your code, I've learned a lot!

**Pull request merged**